net-libs/libmicrohttpd/libmicrohttpd-0.9.77-r1.ebuild

```diff
@@ -0,0 +1,82 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI="8"
+
+inherit linux-info multilib-minimal verify-sig
+
+MY_P="${P/_/}"
+
+DESCRIPTION="Small C library to run an HTTP server as part of another application"
+HOMEPAGE="https://www.gnu.org/software/libmicrohttpd/"
+SRC_URI="mirror://gnu/${PN}/${MY_P}.tar.gz
+	verify-sig? ( mirror://gnu/${PN}/${MY_P}.tar.gz.sig )"
+S="${WORKDIR}"/${MY_P}
+
+LICENSE="|| ( LGPL-2.1+ !ssl? ( GPL-2+-with-eCos-exception-2 ) )"
+SLOT="0/12"
+KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~ia64 ~loong ~m68k ~mips ~ppc ~ppc64 ~riscv ~s390 ~sparc ~x86"
+IUSE="+epoll +eventfd ssl static-libs test +thread-names verify-sig"
+REQUIRED_USE="epoll? ( kernel_linux )"
+RESTRICT="!test? ( test )"
+
+KEYRING_VER=201906
```

Comment:
This allow building both old and new libmicrohttpd versions. The library changed the keyring in the past and will change it in future. With most probably API and ABI bump, slotting would be needed for the library. The infrastructure would be already there.

Reply:
Again, what's the problem with adding new keys to the old ones?

Reply:
If far as I see it is not common way for other packages.
Not sure how to implement it:
* Download individual signing keys (and cache them on mirrors) and combine to the local keyring, add new signing keys when necessary. This way the old keys may be left only on gentoo mirrors, so users must rely on mirrors. Problematic to maintain as when cleanup of old keys would be required (after removal of the old packages versions) there are not clear connections between the keys and the releases. Anyway. it would need individual updates for every keychain update.
* Download the current keyring from the upstream (and cache it on mirrors) and combine it with the old ones. The problems are similar to previous version, but the old key would not be available for sure. To cache the keyring on mirrors, manual update of ebuild would be needed for every key update anyway.
* Download the current keyring during the emerge, do not cache file on mirrors. Non-standard way, typically used on portage only for "live" builds. The new keys would be added to the local installed keyring. In such case the local keyring will grow indefinitely as now cleanup is exist for it. New users will be unable to build the old version of the package as old keys are not available anymore as the "current upstream keyring".
In any such case the signature checking is weaker as it would succeed if outdated old key (stolen, revoked) would be used to sign the new (unmatched) release.

The suggest solution is simple and straightforward.

The packages have versions and the relevant signature keys have versions as well. What's the problem with it?
The world is not frozen, including the keyrings used for signing.

The gentoo itself has keyring versioning: https://github.com/gentoo/gentoo/tree/master/sec-keys/openpgp-keys-gentoo-developers

Reply:
The keys should be added **and removed** from time to time. The reason for the keychain update is not only the new developer joined, but also developers left or keys expired and replaced.

Reply:
It is not an exception that package have versions of the keyring:
https://github.com/gentoo/gentoo/tree/master/sec-keys/openpgp-keys-apache-tomcat
https://github.com/gentoo/gentoo/tree/master/sec-keys/openpgp-keys-bcl
https://github.com/gentoo/gentoo/tree/master/sec-keys/openpgp-keys-bradking
and others

Reply:
I don't understand the point you're trying to make? Literally every keyring package has versioned ebuilds. No one is suggesting to not have a versioned keyring ebuild. @mgorny is questioning why you need to have a special variable to define which version of the keyring you depend on, instead of simply depending on the latest version of the package by default.

If the keyring package adds an additional allowed signing key I'm not sure it's a reasonable vulnerability to worry that a newly added signing key can start signing old releases without authorization.

If the keyring package removes an allowed signing key then the ebuild for the newer release can use a >= requirement on the keyring package. But depending on the reason for removal it might be necessary to pull old releases! Which would mean updating those ebuilds anyway...

This tends to be a rarer occurrence anyway, and it's still possible to update the older ebuilds to add a <= on the keyring package build dependency. It is still only relevant for people doing fresh builds of the old package.

Reply:
Your examples do however demonstrate key files being downloaded via SRC_URI rather than being checked into gentoo.git and referenced from the filesdir.